### ISSUE TITLE
Ontology export columns are incomplete 

### ISSUE THREAD

**Issue by lgeistlinger** [waldronlab/BugSigDB#92]
@tosfos I'd like to filter signatures exported from BugSigDB by review status for #80.
This seems currently not possible as the only column that we have in the export csv files on review status is `Revision editor`.

However, the field `Revision editor` has value "WikiWorks743" for reviewed content (see eg https://bugsigdb.org/Study_255) as well as content that still needs to be reviewed (see eg https://bugsigdb.org/Study_400). Can this be changed so that this column is empty or NA in the export for content that still need to be reviewed?

**Comment by tosfos** [comment]
That sounds good. Should we also add code for the wiki to ignore the WikiWorks... users for these fields? Usually our edits are not very scientifically useful so we shouldn't be getting credit :).

**Comment by lwaldron** [comment]
Yeah that would make sense to hide the Wikiworks74 etc curations :)

**Comment by lgeistlinger** [comment]
There are some more considerations to this though. From the discussion arising in #93, it becomes clear that we would actually like the exported column "Revision editor" to list the person who marked the content as reviewed as @ftzohra22 points out in https://github.com/waldronlab/BugSigDB/issues/93#issuecomment-880132766.

**Comment by lwaldron** [comment]
It seems "Revision editor" is correct, but "Reviewer" could be another column in the export.

**Comment by lgeistlinger** [comment]
Agreed.

**Comment by lwaldron** [comment]
Adding a priority label to this as we will make our first versioned data release (https://github.com/waldronlab/BugSigDBExports/issues/4) as soon as it is resolved.

**Comment by lgeistlinger** [comment]
Moving @lwaldron's comment from #93 here:
> I also noted that the "Complete/Incomplete" status would be worth exporting too. So the two items here are to include in the export
> 
>     1. "Reviewed by" (missing or who marked study/experiment/signature as reviewed), and
> 
>     2. Complete or Incomplete (or TRUE/FALSE)

**Comment by lgeistlinger** [comment]
I'd like to add two more columns in the export:

1. The EFO ID of the corresponding condition investigated
2. The UBERON ID of the corresponding body site sampled

as it arises from https://github.com/waldronlab/BugSigDB/issues/55#issuecomment-877253482.
This will allow for systematic analysis and ontology-based queries downstream of the export in eg [bugsigdbr](https://github.com/waldronlab/bugsigdbr) and/or [BugSigDBStats](https://github.com/waldronlab/BugSigDBStats).

**Comment by tosfos** [comment]
> It seems "Revision editor" is correct, but "Reviewer" could be another column in the export.

Sorry. I'm not clear on how to proceed with this. Right now Curator and Revision Editor can list a WikiWorks user. Do we want to change either one or both?

**Comment by tosfos** [comment]
> It seems "Revision editor" is correct, but "Reviewer" could be another column in the export.

Do we also want the "Reviewer" field to be prominently displayed next to the Revision Editor field (on the site itself)? Or is OK that you need to hover over the (!) icon in order to see it?

**Comment by lwaldron** [comment]
> Sorry. I'm not clear on how to proceed with the WikiWorks user removal. Right now Curator and Revision Editor can list a WikiWorks user. Do we want to change either one or both? And if we are removing the WikiWorks user, is it OK if we end up with these fields as blank?

I don't feel strongly about this, but blank seems equally informative as WikiWorks user, so fine with me. I would want to search for those pages where both curation and revision editor are blank, and at least make sure they are at least reviewed. I don't think it seems worthwhile to dig out again the pre-wiki curators of these pages. 

> Do we also want the "Reviewer" field to be prominently displayed next to the Revision Editor field (on the site itself)? Or is OK that you need to hover over the (!) icon in order to see it?

Again I don't feel strongly about this. Both seem OK, although displaying the Reviewer may be a bit better to recognize reviewing as an important contribution. FYI in case this helps inform a decision: we currently have many unreviewed pages, and although we will catch up some, reviewing is almost as much work as the original curation and it seems likely that curation will always outpace reviewing.

**Comment by tosfos** [comment]
> Complete or Incomplete (or TRUE/FALSE)

This part is now "Complete". The requested column can be found [in any CSV](https://bugsigdb.org/Help:Export).

**Comment by tosfos** [comment]
The 

> It seems "Revision editor" is correct, but "Reviewer" could be another column in the export.

Done.

**Comment by tosfos** [comment]
> I'd like to add two more columns in the export:
> 
>     1. The EFO ID of the corresponding condition investigated (such as [EFO:0001075](http://www.ebi.ac.uk/efo/EFO_0001075) for "ovarian carcinoma")
> 
>     2. The UBERON ID of the corresponding body site sampled (such as [UBERON:0000341](http://purl.obolibrary.org/obo/UBERON_0000341) for "throat")

Please see if this looks OK in the new CSVs.

**Comment by lgeistlinger** [comment]
As noted in #94, it is problematic for downstream applications that these changes to the export files cause the link location to change. It would be preferable if we would have links such as eg `https://bugsigdb.org/export/experiments.csv` that would always point to the latest version of these bulk export files.

**Comment by lgeistlinger** [comment]
Columns `State` (for completion status) and `Reviewer` (for review status) look great.

**Comment by lgeistlinger** [comment]
EFO / UBERON ID: It looks like https://github.com/waldronlab/BugSigDB/issues/55#issuecomment-877253482 precedes this, as these columns are currently mostly blank. We are missing eg the corresponding EFO ID for terms like "adenoma". This is true for the export files but also the pages itself (eg https://bugsigdb.org/Study_1/Experiment_1), where links to the EFO IDs / EFO pages are missing.

Also,  EFO IDs such as `EFO:0001075` and `UBERON:0000341` are currently abbreviated as `1075` and `341` in the export columns, but it would be preferable to have them exported fully spelled out I think.

**Comment by tosfos** [comment]
I figured out what's happening. The leading zeros are being exported, as you can see in a text editor:
![image](https://user-images.githubusercontent.com/10228795/129932311-cc3d8495-002f-4cfb-b900-487538d50a74.png)

The spreadsheet application is interpreting these as numbers and removing the leading zeros, but it's not the application's fault. Semantic MediaWiki should be surrounding these fields with quotes, since they are stored internally as a text data type. This is likely a bug with Semantic MediaWiki. We'll research this and see what we can do.

**Comment by tosfos** [comment]
My previous comment was incorrect. Semantic MediaWiki is doing everything right. The SMW CSV exporter simply uses the PHP native fputcsv to create its CSV files. And the CSV spec appears to say that numbers don't need to be wrapped in quotes. So the spreadsheet application is what is at fault since technically all CSV fields should be interpreted as text, but I guess it's doing its best to figure out what a user would expect.

If we want to work around this we would need to modify the extension to use our own custom CSV encoder, which would be a bit of a project and probably not a good idea.

**Comment by lgeistlinger** [comment]
Right, the IDs are indeed exported with leading zeros (my bad), but would it be possible to export them with leading "EFO:" for condition and leading "UBERON:" for body site? Not a problem if not, as we can also add them downstream in our R application. It would just be more convenient and more straightforward to use for everyone who doesn't use the R application but rather works on the exported files themselves.

**Comment by lwaldron** [comment]
"EFO:" and "UBERON:" are also not a bad idea to include if it's not difficult because there are also other potentially relevant ontologies. We could in theory mix ontologies and use downstream tools to map them, and even if we never do that, it's more communicative to someone downloading the file who isn't familiar with its contents.

**Comment by lgeistlinger** [comment]
It's a great point. As we already observed (https://github.com/waldronlab/BugSigDB/issues/55#issuecomment-801474939), EFO actually already *is* an umbrella ontology, meaning not all EFO IDs start with "EFO:", but also "CHEBI:", "Orphanet:", "HP:", and "MONDO:", ...
In numbers only 11,338 out of 27,175 terms in the EFO start with "EFO:". We thus indeed need the prefixes as available from the Term ID <-> Term name mappings.

**Comment by tosfos** [comment]
We combined the two columns as requested:
![image](https://user-images.githubusercontent.com/10228795/130672797-76843278-4ab3-44d3-895e-ef88ee3a4298.png)

Please let me know if this looks oK now.

**Comment by lwaldron** [comment]
Looks good! My only question is why so many EFO IDs are missing, since all
the condition terms were taken from the EFO?--

Levi Waldron

Associate Professor

Department of Epidemiology and Biostatistics

CUNY Graduate School of Public Health and Health Policy

Institute for Implementation Science in Population Health

55 W 125th St, New York NY 10035

https://waldronlab.io

Join the microbiome Virtual International Forum: https://microbiome-vif.org

**Comment by ftzohra22** [comment]
I also noticed these on CSV exports (my filters were 'streptococcus' 'homo sapiens' and 'feces') :

1. Some conditions and EFO missing as @lwaldron mentioned: My guess is, this could be because a lot of them aren't entered correctly (not validated terms), for example: (https://bugsigdb.org/Study_426) 'Autoimmune Hepatitis' entered as lower case 'autoimmune hepatitis', might have resulted in empty export cells for both condition and EFO.  
2. It looks like this is also the case for the matched on column- 'sex' is missing on the csv which isn't listed as a validated term on the wiki. 

![image](https://user-images.githubusercontent.com/45443641/130762591-147344c2-8a99-4020-94f4-dbd4aa9da307.png)

**Comment by lgeistlinger** [comment]
Two points here:

1. @tosfos: where are the mappings term name (eg "obesity") -> term ID (eg "EFO:00...") are coming from?
2. @ftzohra22 @rimsha1 @lwaldron : we need reviewers to continously screen for invalid entries entered by curators in the condition, body site, matched on, confounders controlled for fields. The clean-up pages are supposed to assist with that.

**Comment by tosfos** [comment]
They come from [the glossary](https://bugsigdb.org/Category:Glossary). Right now, we'd need to create a new page for each condition (although we could import a CSV). However, this also depends on #89. If we go ahead with that, we'll be able to query their API and remove the need for the Glossary.

**Comment by lwaldron** [comment]
I think this is higher priority than #89 and needed for basic functionality, whereas #89 is more of a wishlist. For now, the following would suffice for basic functionality:
1. a page for each condition
2. the ontology columns being complete in the export
3. ideally, the complete EFO and Uberon ontologies in the system for autocomplete etc, but we could live longer with having to enter manually on an as-needed basis like we have been.

**Comment by lwaldron** [comment]
Adding the bug label since specifically for the incomplete ontology columns in the export. That is the final remaining item for our first "official" export of bugsigdb, see https://github.com/waldronlab/BugSigDBExports/issues/4#issuecomment-917712225. This is a high priority item.

**Comment by lgeistlinger** [comment]
Hi @tosfos: how do you think we should best proceed for this? Should I provide a csv for importing or do you go ahead with pulling the mappings directly from EFO and UBERON?

**Comment by tosfos** [comment]
Sorry about the delay! We'll send over a CSV shortly with a header row and an example row. That will help you fill it out.

**Comment by tosfos** [comment]
[Glossary Import Format.csv](https://github.com/waldronlab/BugSigDB/files/7172033/Glossary.Import.Format.csv)
Please use the attached CSV. Here are some notes on the fields:

1. **Title** will be used as the page name
2. **Term** should be lowercase (or an abbreviation), so that we can match it up with the current Body site and Condition fields.
4. **Definition** should be brief and in plain text.
6. **Alias** can accept multiple values, delimited with a semicolon ;
8. **Link** must be a valid URL pointing to UBERON or EFO (for the Body site or Condition terms respectively - see the Oral gland example).
10. **Latin** is a single latin equivalent for the term.

**Comment by lgeistlinger** [comment]
Sounds good. Which field is used to store the term ID (eg "EFO:0001075") for a given term name (eg "ovarian carcinoma")?

**Comment by tosfos** [comment]
We should be able to derive this automatically from the URL. I don't think we need to worry about that field.

**Comment by lgeistlinger** [comment]
Alright. A look at what information is typically available for a term in the official [EFO OBO release](https://www.ebi.ac.uk/efo/efo.obo) - here for term "ovarian carcinoma":

```
[Term]
id: EFO:0001075
name: ovarian carcinoma
def: "A malignant neoplasm originating from the surface ovarian epithelium. It accounts for the greatest number of deaths from malignancies of the female genital tract and is the fifth leading cause of cancer fatalities in women. It is predominantly a disease of older white women of northern European extraction, but it is seen in all ages and ethnic groups. Adenocarcinomas constitute the vast majority of ovarian carcinomas. The pattern of metastatic spread in ovarian carcinoma is similar regardless of the microscopic type. The most common sites of involvement are the contralateral ovary, peritoneal cavity, para-aortic and pelvic lymph nodes, and liver. Lung and pleura are the most common sites of extra-abdominal spread. The primary form of therapy is surgical. The overall prognosis of ovarian carcinoma remains poor, a direct result of its rapid growth rate and the lack of early symptoms. --2002" [NCIT:C4908]
comment: Editor note: unclear why this is distinct from malignant ovarian epithelial tumor in NCIT.
synonym: "carcinoma of ovary" EXACT [MONDO:patterns/carcinoma, NCIT:C4908]
synonym: "carcinoma of the ovary" EXACT [NCIT:C4908]
synonym: "epithelial ovarian cancer" EXACT [NCIT:C4908]
synonym: "ovarian cancer" EXACT [NCIT:C4908]
synonym: "ovarian carcinoma" EXACT [] {comment="Mondo preferred label 15.08.2021."}
synonym: "ovarian carcinoma" EXACT [DOID:4001, NCIT:C4908]
synonym: "ovarian carcinoma" EXACT [] {comment="preferred label from MONDO"}
synonym: "ovarian epithelial cancer" EXACT [NCIT:C4908]
synonym: "ovary carcinoma" EXACT [MONDO:patterns/location]
xref: DOID:4001 {source="EFO:0001075", source="MONDO:equivalentTo"}
xref: EFO:0001075 {source="DOID:4001", source="MONDO:equivalentTo"}
xref: MONDO:0005140
xref: NCIT:C4908 {source="DOID:4001", source="EFO:0001075", source="MONDO:equivalentTo"}
is_a: Orphanet:398934 {source="DOID:4001", source="MONDO:Redundant", source="MONDOLEX:0005140", source="NCIT:C4908"} ! Malignant epithelial tumor of ovary
relationship: EFO:0000784 UBERON:0000992 ! has_disease_location ovary
property_value: closeMatch http://linkedlifedata.com/resource/umls/id/C0677886
property_value: closeMatch http://linkedlifedata.com/resource/umls/id/C0677886
property_value: exactMatch DOID:4001
property_value: exactMatch DOID:4001
property_value: exactMatch NCIT:C4908
property_value: exactMatch NCIT:C4908
property_value: gwas:trait "true" xsd:string
```

That means I will be able to deliver fields 1-5 of your csv with

1. Title -> name
2. Term -> name
3. Definition -> def
4. Alias -> synonym

(lhs: fields in your csv; rhs: fields in the EFO)

5. Links will be of the form:

- http://www.ebi.ac.uk/efo/EFO_0006351 (for term IDs starting with "EFO:")
- http://www.orpha.net/ORDO/Orphanet_137 (for term IDs starting with "Orphanet:")
- http://purl.obolibrary.org/obo/UBERON_0000029 (for all other term IDs, ie term IDs starting with one of {BTO:, CHEBI:, CLO:, CL:, FBbt:, FBdv:, FMA:, GO:, HP:, MONDO:, MP:, NCBITaxon:, OBI:, PATO:, PO:, UBERON:})

(this also provides for how term IDs should be exported in the ontology columns that we request in the thread above).

6. EFO/UBERON do not provide latin equivalents for the terms, and I don't see a way of easily obtaining that otherwise. I will thus leave this field blank if not otherwise indicated.

**Comment by tosfos** [comment]
We added the Latin column since we assumed it would be a nice feature given the site's medical nature. If it won't be used, we'll remove it. Please advise.

On second thought, you can ignore the Term column. It looks like it will always have the same data as the Page Name, so we can just calculate the lowercase of that field instead.

**Comment by lgeistlinger** [comment]
The latin column sounds indeed like a nice feature, but as EFO/UBERON do not provide latin equivalents for the terms, and I don't see a way of easily obtaining that otherwise, I suggest to just drop that column.

I have accordingly prepared the csvs for:

- [Condition](https://www.dropbox.com/s/gcr7ttd5hune4rj/condition_glossary.csv?dl=0) (28,384 terms), and
- [Body site](https://www.dropbox.com/s/pooi3am3h8hkeke/body_site_glossary.csv?dl=0) (14,941 terms) 

I expect some hiccups and some manual work when matching those against the terms that are already on bugsigdb.org. 
The question is whether it's better to clean up offending terms before or after the import (ie those terms that are already on bugsigdb.org but don't have an entry in the import files above).    

And just to keep in the back of our heads: EFO and UBERON update on a regular basis, ie we should start thinking at some point about how to automatically keep bugsigdb.org's glossary in sync with EFO/UBERON - but it is my understanding that this will be part of the postponed activities planned in #89.

**Comment by tosfos** [comment]
Thanks. We'll drop the latin column.

Does it make sense to first clean up some of the strange formatting like:
`293 cell$;A-293 cell$;A293 cell$;HEK cell$;HEK293 cell$;human embryonal kidney cell$;human embryonic kidney cell`

Or is that something we would do after the import?

> some manual work when matching those against the terms

If they are valid terms, it makes sense to just add these as aliases, and these can be discovered and fixed after the import. If they are not valid (like a misspelling), they should probably be cleaned up before the import.

> it is my understanding that this will be part of the postponed activities planned in #89.

Correct.

**Comment by lgeistlinger** [comment]
As described [previously](https://github.com/waldronlab/BugSigDB/issues/55#issuecomment-834696762), the synonyms for a specific term can contain all kind of special characters so some improvisations was needed as to what to use as separating synonyms..
This is technically a comma-separated list of synonyms for each term, but as commas and semi-colons are partly contained in the synonyms - I used `$;` to separate individual synonyms within the list. 
Would this be a separator that you can process or would you need a different separator? (which one)

**Comment by lgeistlinger** [comment]
I could potentially also just replace any existing commas in the synonyms, and provide a comma-separated list of synonyms for each term if that makes processing easier.

**Comment by lgeistlinger** [comment]
> If they are valid terms, it makes sense to just add these as aliases, and these can be discovered and fixed after the import. If they are not valid (like a misspelling), they should probably be cleaned up before the import.

@tosfos I have accordingly updated the csvs for:

- [Condition](https://www.dropbox.com/s/gcr7ttd5hune4rj/condition_glossary.csv?dl=0) (28,391 terms), and
- [Body site](https://www.dropbox.com/s/pooi3am3h8hkeke/body_site_glossary.csv?dl=0) (14,941 terms) 

This required action on the following items on bugsigdb.org for "conditon" - which were not a term title in the efo but either:

(i) represented existing synonyms for an EFO term (eg. "race" for EFO term "ethnic group"), 
(ii) were added as synonyms to appropriate EFO terms where indicated (eg. "sessile serrated adenoma" for EFO term "Colon Sessile Serrated Adenoma/Polyp"), 
(iii) were added as independent condition terms not in EFO but another ontology such as CHEBI (eg "medroxyprogesterone acetate"),
(iv) or were cleaned up on bugsigdb.org itself (eg. "irritable bowel sydrome" -> "irritable bowel sy**n**drome").

"sessile serrated adenoma"
  "oral halitosis"    
  "sjogren's syndrome"
  "substance related disorder"    
  "titanium dioxide nanoparticles"
  "air pollution"    
  "crohn disease"
  "medroxyprogesterone acetate"    
  "cervical cerclage"
  "equol"    
  "irritable bowel sydrome"
  "urinary track infection"    
  "kidney stone"    
  "female reproductive organ cancer"
  "end stage renal disease"    
  "hiv/aids pre-exposure prophylaxis"
  "race"    
  "antibiotic"    

No such actions were needed for body site / UBERON.

> I could potentially also just replace any existing commas in the synonyms, and provide a comma-separated list of synonyms for each term if that makes processing easier.

I discarded this idea, as a quick check of synonyms in EFO revealed a total of 14,344 synonyms containing a comma. This includes numerous instances where replacing a comma would obstruct meaning as eg for chemical notations such as "(3S)-3-(4-hydroxyphenyl)-3,4-dihydro-2H-1-benzopyran-7-ol". I thus suggest going ahead with "$;" as separator in the alias/synonym field if that can be processed on your end.

Both files would thus be ready for import on my end.

**Comment by lwaldron** [comment]
Great, @lgeistlinger. From my perspective, #73 getting the dynamic home page onto the home page, followed closely by this issue, are our top priorities.

**Comment by tosfos** [comment]
We tested out the "$;" as a delimiter and it seems to work fine. We did a test import of some of these terms, and you can see 17 new terms as part of the [Glossary](https://bugsigdb.org/Category:Glossary). 

Questions about the aliases.

1. When using the form, it will only autocomplete based on the Condition list that is enumerated on the [Condition property page](https://bugsigdb.org/Property:Condition). It will not offer any aliases. Is that an issue?
2. Similarly, if an alias has been added as the value for one of these fields, the page will work that it is now an allowed value for this property. Is that an issue?
3. Is there any particular way that you're expecting the wiki to treat aliases? For example, I'm thinking that if a Condition is set to an alias, instead of semantically storing that alias as this Experiment's Condition/Body site, we could store the "main" glossary term that corresponds to that alias. We could also **display** the "main" term on the Experiment page, though that might be confusing when editing the page. (This would solve the issue mentioned in item 2.)

**Comment by lwaldron** [comment]
1. That seems fine to me. 
2. That seems OK, although it would be nice to have a method for cleanup of the Condition property page. i.e. I noticed we currently have:
    - nicotine dependence
    - Nicotine dependence
    - non-alcoholic fatty liver
    - non-alcoholic fatty liver disease
I noticed these by eye but it would be great to see any unknown Conditions here and be able to tell administrators / reviewers how to avoid entering or at least identify non-standard, ie non-ontology main terms. 
3. That sounds nice. In an ideal world it would not matter whether someone entered or searched for a term or its alias, because they would be equivalent but the term's page would belong to the main term.  However I'm not even sure whether the aliases of distinct terms are always disjoint, if not, that may not be possible anyways. This is not critical but some redirecting of aliases to their "main" glossary term would be very nice.

**Comment by tosfos** [comment]
> I noticed these by eye but it would be great to see any unknown Conditions here and be able to tell administrators / reviewers how to avoid entering or at least identify non-standard, ie non-ontology main terms.

Are you referring to terms that are neither "main" terms nor aliases? If so, we already track fields that are set improperly.

**Comment by lgeistlinger** [comment]
Just wanted to quickly check in in here one more time as to whether the Oct 25 deadline is realistic for that? It's no problem at all if don't make it, then we introduce changes somewhere along the way, but it would make for a nice clean introduction into the Bioc release if we were able to meet that date. Thanks!

**Comment by tosfos** [comment]
We did another test import. Please see an example Condition page [here](https://bugsigdb.org/HEK-293_cell). 

You can see a form that has an Alias value entered for a Body site. See the "body site" field [here](https://bugsigdb.org/w/index.php?title=Study_24/Experiment_1&action=formedit), which is set to "salivary gland". And you can see that the resulting page shows the "main" term instead of the alias, [here](https://bugsigdb.org/w/index.php?title=Study_24/Experiment_1&action=formedit), so it shows "oral gland ".

**Comment by tosfos** [comment]
The import is now complete. Please note that we weren't able to import these Aliases since there were special characters (brackets) in the titles:

> [Pyruvate dehydrogenase acetyl-transferring] kinase isozyme 1, mitochondrial measurement
> Alcohol dehydrogenase [NADP+] measurement
> Superoxide dismutase [Mn], mitochondrial measurement
> O-[(4Z)-decenoyl]carnitine
> O-[(9Z)-tetradecenoyl]-L-carnitine
> 1,2-di-[(9Z,12Z)-octadecadienoyl]-sn-glycero-3-phosphocholine
> (1->3)-beta-D-glucan
> S-[2,3-bis(palmitoyloxy)propyl]-Cys-Ser-Lys-Lys-Lys-Lys
> 1-[(4Z,7Z,10Z,13Z,16Z)-docosapentaenoyl]-sn-glycero-3-phosphocholine
> 1-tetradecanoyl-2-[(9Z,12Z)-octadecadienoyl]-sn-glycero-3-phosphocholine
> 1-hexadecanoyl-2-[(7Z,10Z,13Z,16Z)-docosatetraenoyl]-sn-glycero-3-phosphocholine
> 4-(4-methylpiperazin-1-yl)-7-(trifluoromethyl)pyrrolo[1,2-a]quinoxaline dimaleate
> 3-[4-(3-chlorophenyl)piperazin-1-yl]-1,1-diphenylpropan-2-ol hydrochloride
> 1-[2-(benzhydryloxy)ethyl]-4-(3-phenylpropyl)piperazine dihydrochloride
> 1-[1-(1-benzothiophen-2-yl)cyclohexyl]piperidine maleate

**Comment by tosfos** [comment]
We also modified the layout of Glossary pages as requested. You can see the results [here](https://bugsigdb.org/(%2B)-abscisic_acid).

**Comment by lwaldron** [comment]
Looks great, @tosfos . One side-effect that I notice is that these glossary pages dominate the simple search, e.g. see below for "cancer", of which some do and some don't appear in any Experiment. This might be disappointing to users when the glossary pages don't link further back to Study or Experiment pages (even in cases that that condition does appear in some experiments), leading them to think search results aren't relevant to bugsigdb.org. A couple possible workarounds I could think of are 1) exclude glossary terms from the simple search results, 2) use a semantic search (as in the drilldown) also in the simple search box. Of course this is not really related to the current issue and I'm sure this belongs in a separate issue...

![image](https://user-images.githubusercontent.com/639942/140535727-f97188e9-de0d-4ea2-a189-f49e2e1430cd.png)

**Comment by lgeistlinger** [comment]
> The import is now complete, as you can see here.

@tosfos That is great! Also amazing how one can now use general search for getting to the corresponding glossary page!

Now the main point of this issue here is whether this information gets also properly exported. That doesn't seem to be the case if I take a look at the exported experiments.csv file, which still contains a lot of blank fields for valid glossary terms such as eg [adenoma](https://bugsigdb.org/Adenoma) or [meconiumn](https://bugsigdb.org/Meconium). 

<img width="960" alt="experiments_exported" src="https://user-images.githubusercontent.com/11871050/140567296-acb3ec53-177a-4f0a-bf03-431f0b5cc2c6.png">

**Comment by tosfos** [comment]
> these glossary pages dominate the simple search

There's no clear answer to how to work around this. I created the separate #105 .

**Comment by tosfos** [comment]
> Now the main point of this issue here is whether this information gets also properly exported.

We haven't switched to export to use the new data. Will do.

**Comment by tosfos** [comment]
We noticed some language markers on pages like [this](https://bugsigdb.org/w/index.php?title=Feces&oldid=38015). We actually can semantically store these terms together with the language code, but I'm not sure where we would take advantage of this data. Is there any intention for special multilingual wiki functionality?

**Comment by lgeistlinger** [comment]
Interesting, I saw that too. I don't think we would have a particular use case for that nor do I expect curators to really make use of it, as I expect all curated articles to be in English.

**Comment by tosfos** [comment]
Right now, we have separate columns for Uberon and EFO IDs. And we have icons for Uberon, EFO and ORPHANET. But I'm seeing URLS for MONDO, OBI, HP and others. How should we handle these? For example, see [here](https://bugsigdb.org/Hereditary_gastric_cancer). Is the icon correct? And which column would we use for the ID? Should there only be one ID column in the export?

**Comment by lgeistlinger** [comment]
Thanks @tosfos. As explained [here](https://github.com/waldronlab/BugSigDB/issues/92#issuecomment-902839324), we'd like to have all IDs imported from the condition.csv to be exported in the `EFO ID` column; and all IDs imported from the bodysite.csv to exported in the bodysite.csv. 

For body site, this will always be UBERON IDs. But for condition this will be all kind of EFO IDs (not all start with "EFO:", but also "MONDO:", "ORPHANET", ... or even "UBERON", as some UBERON terms are part of the EFO) - and those we'd like to export in the EFO ID column. Thanks!

**Comment by tosfos** [comment]
Right. Sorry that I keep getting confused about this.

**Comment by tosfos** [comment]
It seems like certain terms exist in both the Body site and the Condition CSVs. Is that something that should be possible? See, for example, [blood](https://bugsigdb.org/w/index.php?title=Blood&diff=prev&oldid=46502).

**Comment by lgeistlinger** [comment]
Yes, this is correct, there are certain UBERON terms contained in the EFO. We haven't seen examples of curators picking up certain body sites (such as blood) as conditions, but it is technically possible.

<img width="1222" alt="Screen Shot 2021-11-11 at 11 23 00 AM" src="https://user-images.githubusercontent.com/11871050/141332840-8951bbb1-e13a-4ca0-b718-8aa2ace8d7f4.png">
<img width="1296" alt="Screen Shot 2021-11-11 at 11 23 27 AM" src="https://user-images.githubusercontent.com/11871050/141332861-ae1579f2-8ddb-46d9-a198-5e788e3ff168.png">

**Comment by tosfos** [comment]
Got it. There is an issue then on the MediaWiki side. We can't have 2 different pages that share the same title. And if we try to manufacture 2 different titles like "Blood (condition)" and "Blood (body site)" it makes it more complicated to match the correct Glossary page with the correct term (which is just entered as "Blood").

Would it make sense that the Condition and Body site definitions of a single term could be different? Or would have 2 different URLs? Or would the "Condition" and "Body site" definitions always be the same? If the 2 Glossary pages would be alike in every way, what we can do is allow a single Glossary page to belong to both categories.

**Comment by lgeistlinger** [comment]
I think if this simplifies things, we can just remove the UBERON terms from the condition.csv - as there use for condition is largely hypothetical and not something that we've observed in practice when curating > 500 articles.

**Comment by lgeistlinger** [comment]
> Or would the "Condition" and "Body site" definitions always be the same? If the 2 Glossary pages would be alike in every way, what we can do is allow a single Glossary page to belong to both categories.

I think that would also be a perfectly fine solution.

**Comment by tosfos** [comment]
Some more questions regarding the aliases. 

If someone enters an alias, the wiki tries to find the Main Term (Glossary page title) in the Glossary. That is, a Glossary term that has this alias listed. Once a match is found, it displays the Main Term instead of the Alias that was entered. It also finds ALL of the aliases for this main term and inserts them in the page (but hides them) so that a search for any alias of this term will find this page.

A similar behavior applies if someone actually enters a term that IS a Main Term in the Glossary. The wiki will display this Main Term but also look up all aliases for this term and stores it in the page.

Please let me know if any of the above is incorrect behavior.

We are running into an issue in that sometimes the same alias is listed on more than one Glossary page. For example, [gastric carcinoma](https://bugsigdb.org/Gastric_carcinoma) and [stomach neoplasm](https://bugsigdb.org/Stomach_neoplasm) both list "Gastric Cancer" as an alias. And to complicate things, gastric cancer is also a [Glossary Main Term with its own page](https://bugsigdb.org/Gastric_cancer).

What should the wiki do in this situation? Right now it tries to find EVERY term that matches "Gastric Cancer" and it lists all 3 of them on the page like:

Condition: gastric cancer , gastric carcinoma , stomach neoplasm 

And it also stores the full list of all aliases from all 3 Glossary pages within the page. Is this correct?

I'm thinking it would be better to show the Main Term only. And, if a matching Main Term is not found, then only show the Main Term of the first Glossary page that has a matching alias. But under all circumstances only one Condition (or Body Site) should be displayed. What do you think?

And also, how should we deal with storing aliases then? Do we want to cast the net as wide as possible and find every single matching Alias? Or just the aliases of the first matching term that was found? (Note that there is nothing special about the first matching term. It's probably the first one the wiki will find, alphabetically.)

**Comment by lgeistlinger** [comment]
I believe we should follow here the behavior on EFO site itself, and push curators upon entering an alias into the condition form towards selecting a main term. That means if a curator enters a synonym, autocomplete options presented would be corresponding main terms for that alias only. (That means I would here indeed depart from allowing curators to be able to enter an alias).

Using the "gastric cancer" example, and looking into which autocomplete options become available when typing "gastric cancer" into the search field, it seems that EFO provides corresponding main terms (potentially mapped via synonyms as for "stomach neoplasm"). 

<img width="1266" alt="Screen Shot 2021-11-14 at 6 49 37 PM" src="https://user-images.githubusercontent.com/11871050/141704215-5a895680-0629-41b9-b66a-50857c34a539.png">

This is also true when specifically typing "stomach cancer", which is an alias for "gastric cancer" (and potentially also others). 
 
<img width="1476" alt="Screen Shot 2021-11-14 at 6 50 34 PM" src="https://user-images.githubusercontent.com/11871050/141704107-fc8f75b3-e9ed-4e5d-a0cf-e3ffdea4b257.png">

**Comment by tosfos** [comment]
Interesting idea. Let me see what we can do.

But we already have the pages we imported which already contain aliases. How should we deal with these existing pages?

**Comment by lgeistlinger** [comment]
> But we already have the pages we imported which already contain aliases. How should we deal with these existing pages?

Sorry for the delay in replying. I am not sure I fully understand the practical implications here. Why would the imported pages pose a problem?

**Comment by lgeistlinger** [comment]
I also just took another look into the exported experiments.csv and was happy to see that the gaps in the EFO ID / UBERON ID columns are getting smaller. However, we seem to still have some issues for valid conditions and body sites that were part of the import.

**Conditions with empy EFO ID column, although they represent valid glossary terms**

[Genital neoplasm, female](https://bugsigdb.org/Genital_neoplasm,_female) (there seems to be hiccup due to the ", " notation, see eg https://bugsigdb.org/Study_305)

**Body sites with empty UBERON ID column, although they represent valid glossary terms**:

[breast](https://bugsigdb.org/Breast) (is turned into "obsolete_mammary gland", see eg https://bugsigdb.org/Study_9)
[lung](https://bugsigdb.org/Lung) (is turned into "lung neoplasm", see eg https://bugsigdb.org/Study_5)
[nasal cavity](https://bugsigdb.org/Nasal_cavity) (is turned into "obsolete_olfactory pit", see eg https://bugsigdb.org/Study_302)
[gingiva](https://bugsigdb.org/Gingiva) (is turned into "obsolete_gum", see eg https://bugsigdb.org/Study_284)
[hypopharynx](https://bugsigdb.org/Hypopharynx) (is turned into "obsolete_laryngopharynx", see eg https://bugsigdb.org/Study_394)
[ovary](https://bugsigdb.org/Ovary) (is turned into "obsolete_animal ovary", see eg https://bugsigdb.org/Study_517)

P.S.: Editing conditions and body sites work great, but autocomplete options take a while to load. Any chance this can be faster than that?

**Comment by tosfos** [comment]
> Genital neoplasm, female (there seems to be hiccup due to the ", " notation, see eg https://bugsigdb.org/Study_305)

Yes, it looks like the comma is the culprit. Will fix.

> Body sites with empty UBERON ID column, although they represent valid glossary terms:

It looks to me like the likely culprit is that the wiki is finding a Glossary term that is listed as a Condition even though this is the Body site field. We'll limit the search and see if that fixes all of these.

> Editing conditions and body sites in the experiment form works great, but autocomplete options take a while to load.

Previously all the values were being loaded at the time the form was first loaded, which meant that autocomplete was very quick. Once we imported the full set of these fields' values, we had to change the autocomplete method we were using. Instead of loading them all at the time the form was being loaded (which would no longer be viable), we're now loading them via AJAX while the user types in each field. It is slow though. Let me see what we can do.

**Comment by tosfos** [comment]
> Why would the imported pages pose a problem?

The wiki treats all data the same, regardless of if it comes from the CSV import, or if it is manually inserted by a user. So for the imported pages, the wiki will perform the same alias lookup. In other words, if a Main term wasn't entered in the CSV, the wiki will search on-the-fly to find a matching Main term that has the entered term as one of its aliases. So what happens if multiple matching Main terms are found that all have the entered term as one of its aliases. We need to figure out which Main term should be displayed and store in the Condition/Body site field. Does that make sense?

**Comment by lgeistlinger** [comment]
Yes that makes sense. Let me start with saying that I don't think showing multiple matching terms is a bad solution. In the end, these are coming in via existing aliases and there are certainly situations, where showing multiple main terms better describes the condition than just one main term. For example, we found "(one sort of) cancer" / "(one sort of) carcinoma" being used rather exchangably in the literature, and instead of having two sets of signatures (ie one for "(one sort of) cancer" and one for "(one sort of) carcinoma"), it simplifies downstream analysis if those come grouped together right from the start.

That said, do you internally distinguish between main terms and aliases? Or do you throw them all into one pot when comparing against the conditions / body sites that were entered by the curators into the experiment forms? If you distinguish, one solution might be to first check whether there is a main term that **exactly** matches the condition / body site entered by a curator on the experiment form. If so, then take the main term. If there is no exactly matching main term, go the alias route, and here there might be indeed multiple matching main terms, and here there will be no best solution, and showing all would be just fine I think.

Taking the example of "urinary tract infection' as in https://github.com/waldronlab/BugSigDB/issues/107; here we have an exact matching main term [urinary tract infection](https://bugsigdb.org/Urinary_tract_infection) which could be displayed. There is not necessarily a need for going down the alias route and also bringing in "bacterial urinary tract infection".

**Comment by tosfos** [comment]
> Yes, it looks like the comma is the culprit. Will fix.

This should [now be fixed](https://bugsigdb.org/Study_305).

**Comment by tosfos** [comment]
> That said, do you internally distinguish between main terms and aliases? Or do you throw them all into one pot when comparing against the conditions / body sites that were entered by the curators into the experiment forms?

Right now I don't think there is any internal difference between entered main terms and aliases. If an alias is entered, we just store the main term anyway.

How should we proceed then? (It sounded like your instructions were only intended if we did distinguish.)

**Comment by lgeistlinger** [comment]
I think if the [aforementioned issues](https://github.com/waldronlab/BugSigDB/issues/92#issuecomment-979465386) can be fixed, we are actually in a good spot with the solution that you put in place.

**Comment by tosfos** [comment]
Everything look better now. Any filled body site field now has a matching UBERON ID. The only conditions with missing EFO ID columns are:

1. hepatitis, alcoholic
2. antibiotic exposure
3. antibiotic treatment
4. tumor grade
5. viral lung infection
6. gonorrhea
7. healthy
8. perinatal antibiotics

**Comment by tosfos** [comment]
Thanks! Probably just needs a data rebuild after some new changes we made. We'll check.

**Comment by lgeistlinger** [comment]
Great, thanks!

**Comment by lgeistlinger** [comment]
Just did a fresh export of the experiments.csv - and everything looked good, no more missing items! I believe we can close this.

**Comment by tosfos** [comment]
Great!
